README.md
# Dartnet
Configurable http server in dart.

- [ ] For production server
- [ ] For dev server (`pub serve` proxy)

## Install

`pub global activate dartnet`

### Usage
`dartnet`

or

`pub run dartnet:dartnet`

### Commands

```
Usage 'dartnet' :
        -c, --config    (defaults to "dartnet.yaml")
        -h, --help      

COMMANDS:
        init    Create config file with default value.
                -f, --filename    (defaults to "dartnet.yaml")
        dockerize       Create a Dockerfile from the Dartnet config file
                -f, --filename    (defaults to "dartnet.yaml")

```

## Configuration

Dartnet use a configuration file (default: 'dartnet.yaml')

### Basic
```yaml
address: "0.0.0.0"
port: 1337
root_directory: build
```

This will serve all file under the 'build' directory.

### Complete

```yaml
address: "0.0.0.0" #default: 0.0.0.0

port: 1337 #default: 8080

root_directory: web #default: build

multithread: true #default: true

log: info #default: info

log_file: "dartnet.log" #default: dartnet.log

list_directory: false #default: false

gzip: true #default: true

redirections:
  path:
    /: "index.html" #default behavior
    /redirect: "https://www.google.com"
    /**: "index.html"
  404: error.html
    
https:
  cert: ssl/cert.pem
  key: ssl/key.pem
  password_key: "dartnet"
```

## Run example

```
cd example
dartnet
```

CHANGELOG.md
# Changelog

## 0.3.0

- dart 2 sdk

## 0.2.0

- Implement Path redirection
- Config for redirections change, see example in README
- Add `dartnet dockerize` command

## 0.1.2

- update Jaguar ro 0.4.0

## 0.1.1

- `dartnet init` command: create the default config file

## 0.1.0

- Change configuration file, remove "server" key
- Add gzip compression

## 0.0.2

- Add SSL

## 0.0.1

- Initial version, created by Stagehand


lib/src/utils.dart
import 'dart:io';
import 'dart:async';
import 'package:path/path.dart' as p;
import 'package:jaguar/jaguar.dart';
import 'handler.dart';

const Map<String, String> extMapper = const {
  "html": "text/html",
  "js": "application/javascript",
  "css": "text/css",
  "dart": "application/dart",
  "png": "image/png",
  "jpg": "image/jpeg",
  "jpeg": "image/jpeg",
  "gif": "image/gif"
};

String fileType(File file) =>
    extMapper[file.path.split(".").last] ?? "text/plain";

StreamResponse sendFile(File file) => StreamResponse(
      file.openRead(),
      mimeType: MimeTypes.ofFile(file),
    );

void logResponse(
    Response response, Request request, Duration processingDuration) {
  dartnet.log.info(
      "[${request.method}] ${request.uri.path} - ${response.statusCode} - ${processingDuration.inMicroseconds / 1000}ms");
}

Response listDirectory(Directory dir) {
  List<FileSystemEntity> entities = dir.listSync();

  String currentPath = dir.path.replaceFirst(dartnet.rootDirectory.path, "");
  while (currentPath.startsWith("/"))
    currentPath = currentPath.replaceFirst("/", "");
  currentPath = "/$currentPath";

  List<Map> list = [];
  entities.forEach((FileSystemEntity e) {
    list.add({
      "href": '$currentPath${e.path.replaceFirst(dir.path, "")}',
      "text": e.path.replaceFirst(dir.path, "")
    });
  });

  String render =
      dartnet.listTmpl.renderString({"currentPath": currentPath, "list": list});
  return new Response(render)..headers.mimeType = "text/html";
}

String pathFromRootDir(String path) => "${dartnet.rootDirectory.path}/$path";

Future<Response> findEntity(String path) async {
  final entity = File(pathFromRootDir(path));
  final exist = await entity.exists();
  if (exist == true) {
    return fileSystemToResponse(entity);
  } else if (dartnet.listDirectory == true) {
    final dir = Directory(pathFromRootDir(path));
    final dirExist = await dir.exists();
    if (dirExist) {
      return fileSystemToResponse(entity);
    }
  }
  return null;
}

FileSystemEntity pathInCache(String path) => dartnet.cache[path];

Future<Response> onNotFound() async {
  final path = dartnet.redirections.onNotFound;
  if (path != null) {
    final file = File(pathFromRootDir(path));
    final exist = await file.exists();
    if (exist) {
      return fileSystemToResponse(file);
    }
  }
  return null;
}

Response fileSystemToResponse(FileSystemEntity entity) {
  if (entity is File) {
    return sendFile(entity);
  } else if (entity is Directory) {
    return listDirectory(entity);
  }
  return null;
}

Response errorTemplate(int error) {
  String render = dartnet.errorTmpl.renderString({"error": error});
  return new Response(render)
    ..headers.mimeType = "text/html"
    ..statusCode = error;
}

const String proxyName = "Darnet";
Future<Response> responseFromCache(Request request, String path) async {
  if (dartnet.cache[path] != null) {
    if (dartnet.cache[path] is FileSystemEntity) {
      return fileSystemToResponse(dartnet.cache[path]);
    } else if (dartnet.cache[path] is Uri) {
      return redirect(request, path, dartnet.cache[path]);
    }
  }
  return null;
}

HttpClient _client = new HttpClient();

Future<Response> redirect(Request req, String from, Uri to) async {
  Uri requestUrl = to.resolve(from);
  _client.badCertificateCallback = (_, __, ___) {
    return true;
  };

  final HttpClientRequest clientReq =
      await _client.openUrl(req.method, requestUrl);

  req.headers.forEach((String key, dynamic val) {
    clientReq.headers.add(key, val);
  });
  clientReq.headers.set(HttpHeaders.HOST, requestUrl.authority);

  // Add a Via header. See
  // http://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.45
  clientReq.headers.add(HttpHeaders.VIA, '${req.protocolVersion} $proxyName');

  clientReq.add(await req.body);
  final HttpClientResponse clientResp = await clientReq.close();

  final servResp = new Response<Stream<List<int>>>(clientResp,
      statusCode: clientResp.statusCode);

  clientResp.headers.forEach((String key, List<String> vals) {
    servResp.headers.headers[key] ??= [];
    for (String val in vals) {
      if (servResp.headers.headers[key].contains(val) == false) {
        servResp.headers.headers[key].add(val);
      }
    }
  });

  servResp.headers.removeAll('x-content-type-options');
  servResp.headers.removeAll('x-frame-options');
  servResp.headers.removeAll('x-xss-protection');

  // Add a Via header. See
  // http://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.45
  servResp.headers.add(HttpHeaders.VIA, '1.1 $proxyName');

  // Remove the transfer-encoding since the body has already been decoded by
  // [client].
  servResp.headers.removeAll(HttpHeaders.TRANSFER_ENCODING);

  // If the original response was gzipped, it will be decoded by [client]
  // and we'll have no way of knowing its actual content-length.
  if (clientResp.headers.value(HttpHeaders.CONTENT_ENCODING) == 'gzip') {
    servResp.headers.removeAll(HttpHeaders.CONTENT_ENCODING);
    servResp.headers.removeAll(HttpHeaders.CONTENT_LENGTH);

    // Add a Warning header. See
    // http://www.w3.org/Protocols/rfc2616/rfc2616-sec13.html#sec13.5.2
    servResp.headers.add(HttpHeaders.WARNING, '214 $proxyName "GZIP decoded"');
  }

  // Make sure the Location header is pointing to the proxy server rather
  // than the destination server, if possible.
  if (clientResp.isRedirect &&
      clientResp.headers.value(HttpHeaders.LOCATION) != null) {
    var location = requestUrl
        .resolve(clientResp.headers.value(HttpHeaders.LOCATION))
        .toString();
    if (p.url.isWithin(requestUrl.toString(), location)) {
      servResp.headers.set(HttpHeaders.LOCATION,
          '/' + p.url.relative(location, from: requestUrl.toString()));
    } else {
      servResp.headers.set(HttpHeaders.LOCATION, location);
    }
  }

  return servResp;
}


lib/src/logger.dart
import 'dart:io';
import 'package:logging/logging.dart';
import 'package:ansicolor/ansicolor.dart';

void initLogger(Level logLevel, File logFile) {
  if (logFile?.existsSync() == false) {
    logFile.createSync(recursive: true);
  }
  Logger.root.onRecord.listen((LogRecord record) {
    if (record.loggerName != "J") {
      String log =
          "${record.time} (${record.level.toString()}) ${record.loggerName} -";
      log +=
          " ${record.error ?? ""}${record.message ?? ""}${record.stackTrace ?? ""}";
      if (record.level >= logLevel) {
        print(log);
      }
      logFile?.writeAsString("$log\n", mode: FileMode.APPEND);
    }
  });
}

const Map<String, Level> logLevels = const {
  'ALL': Level.ALL,
  'OFF': Level.OFF,
  'FINEST': Level.FINEST,
  'FINER': Level.FINER,
  'FINE': Level.FINE,
  'CONFIG': Level.CONFIG,
  'INFO': Level.INFO,
  'WARNING': Level.WARNING,
  'SEVERE': Level.SEVERE,
  'SHOUT': Level.SHOUT
};

final AnsiPen red = new AnsiPen()..red(bold: true);
final AnsiPen green = new AnsiPen()..green(bold: true);
final AnsiPen yellow = new AnsiPen()..yellow(bold: true);
final AnsiPen blue = new AnsiPen()..blue(bold: true);


lib/src/server.dart
library dartnet;

import 'dart:io';
import 'package:jaguar/jaguar.dart';

import 'config.dart';
import 'handler.dart';

Future<void> start({String configPath: dartnetConfigurationFile}) async {
  try {
    dartnet = new DartnetConfiguration(configFileName: configPath);
  } catch (e, s) {
    print(e);
    print(s);
    return;
  }

  final server = new Jaguar(
    multiThread: dartnet.isMultithread,
    port: dartnet.port,
    address: dartnet.address,
    securityContext: dartnet.security,
    autoCompress: dartnet.gzip,
  );

  server.get('/*', cacheHandler);

  for (final path in dartnet.redirections.paths) {
    final redirection = dartnet.redirections[path];
    server.get(
      path,
      pathRedirectionHandler(redirection),
    );
  }

  server.get('/', pathRedirectionHandler(Uri.parse('/index.html')));
  server.get('/*', dartnetHandler);

  dartnet.log.info("Start on ${dartnet.address}:${dartnet.port}");
  await server.serve();
}

void initConfigFile({String filename: dartnetConfigurationFile}) {
  File configFile = new File(filename);

  if (configFile.existsSync() == true) {
    stderr.writeln("ERROR: File '$filename' already exist.");
  } else {
    configFile.createSync(recursive: true);
    configFile.writeAsStringSync(_defaultConfig);
  }
}

void dockerize({String filename: dartnetConfigurationFile}) {
  File configFile = new File(filename);

  if (configFile.existsSync() == false) {
    stderr.writeln("ERROR: File '$filename' does not exist.");
  } else {
    DartnetConfiguration config =
        new DartnetConfiguration(configFileName: filename);
    File dockerFile = new File("Dockerfile");
    if (dockerFile.existsSync() == false) {
      dockerFile.createSync();
    }
    dockerFile.writeAsStringSync(
      _dockerFileContent(filename, dartnet.rootDirectoryPath, config.port),
    );
  }
}

String _dockerFileContent(String configFile, String rootDirectory, int port) =>
    'FROM google/dart\n\n'
    'RUN pub global activate dartnet\n'
    'ENV PATH \$PATH:~/.pub-cache/bin\n\n'
    'WORKDIR /dartnet\n\n'
    'RUN cp ~/.pub-cache/bin/dartnet ./\n'
    'ADD $rootDirectory /dartnet/$rootDirectory\n'
    'ADD $configFile /dartnet/\n\n'
    'EXPOSE $port\n\n'
    'ENTRYPOINT ["./dartnet", "-c", "$configFile"]';

String get _defaultConfig => 'address: "0.0.0.0"\n'
    'port: 8080\n'
    'log: "INFO"\n'
    'log_file: "dartnet.log"\n'
    'root_directory: "./"\n'
    'list_directory: false\n'
    'multithread: false\n'
    'gzip: true\n'
    'redirections:\n'
    '  #404: "error.html"\n'
    '  path:\n'
    '    #/: "index.html" #default behavior\n'
    'https:\n'
    '  #cert: "ssl/cert.pem"\n'
    '  #key: "ssl/key.pem"\n'
    '  #password_key: "<PASSWORD>"\n';


lib/src/handler.dart
library dartnet.handler;

import 'dart:async';
import 'package:jaguar/jaguar.dart';
import 'config.dart';
import 'utils.dart';

DartnetConfiguration dartnet;

FutureOr<Response> cacheHandler(Context context) async {
  final timer = Stopwatch()..start();
  final entity = await responseFromCache(context.req, context.req.uri.path);
  if (entity != null) {
    logResponse(entity, context.req, timer.elapsed);
  }
  timer.stop();
  return entity;
}

FutureOr<Response> dartnetHandler(Context context) async {
  final timer = new Stopwatch()..start();

  Response response;

  response = await findEntity(context.req.uri.path);

  response ??= await onNotFound();

  response ??= errorTemplate(404);

  logResponse(response, context.req, timer.elapsed);
  timer.stop();
  return response;
}

RouteHandler pathRedirectionHandler(Uri redirect) => (Context ctx) async {
      Stopwatch timer = new Stopwatch()..start();
      if (redirect.scheme == "http" || redirect.scheme == "https") {
        dartnet.cache[ctx.req.uri.path] = redirect;
      } else {
        dartnet.cache[ctx.req.uri.path] = findEntity(redirect.path);
      }
      final response = await responseFromCache(ctx.req, ctx.req.uri.path);
      if (response != null) {
        logResponse(response, ctx.req, timer.elapsed);
      }
      return response;
    };


lib/src/config.dart
import 'dart:io';
import 'package:yaml/yaml.dart';
import 'package:mustache/mustache.dart';
import 'package:logging/logging.dart';
import 'package:resource/resource.dart';
import 'logger.dart';

const String dartnetConfigurationFile = "dartnet.yaml";
const String dartnetLogFile = "dartnet.log";
const String dartnetLoggerName = "dartnet";

dynamic _getFromMap(Map map, dynamic key) =>
    map != null ? (map[key] ?? null) : null;

class DartnetConfiguration {
  Logger _log;
  Logger get log => _log;

  Map _config;
  RedirectionConfig _redirections;
  HttpsConfig _https;

  Template _listTmpl;
  Template _errorTmpl;

  Template get listTmpl => _listTmpl;
  Template get errorTmpl => _errorTmpl;

  SecurityContext _security;
  SecurityContext get security => _security;

  Directory _rootDirectory;
  Directory get rootDirectory => _rootDirectory;

  final Map<String, dynamic> cache = {};

  DartnetConfiguration({String configFileName: dartnetConfigurationFile}) {
    _log = new Logger(dartnetLoggerName);
    initLogger(logLevel, new File(logFile));

    File configFile = new File(configFileName);
    if (configFile.existsSync() != true) {
      throw "No Configuration File found at '$configFileName'";
    }

    _config = loadYaml(configFile.readAsStringSync());

    _redirections = new RedirectionConfig(
        _getFromMap(_config, RedirectionConfig.redirectionsKey));

    _https = new HttpsConfig(_getFromMap(_config, HttpsConfig.httpsKey));

    if (_https.isValid) {
      _security = new SecurityContext()
        ..useCertificateChain(_https.certPath)
        ..usePrivateKey(_https.keyPath, password: _https.passwordKey);
    }

    _rootDirectory = new Directory(rootDirectoryPath);

    Resource resourceList =
        new Resource("package:dartnet/src/template/list.html");
    resourceList.readAsString().then((template) {
      _listTmpl = new Template(template);
    });
    Resource resource404 =
        new Resource("package:dartnet/src/template/error.html");
    resource404.readAsString().then((template) {
      _errorTmpl = new Template(template);
    });
  }

  static const String portKey = "port";
  static const String addressKey = "address";
  static const String multithreadKey = "multithread";
  static const String rootDirectoryKey = "root_directory";
  static const String logLevelKey = "log";
  static const String logFileKey = "log_file";
  static const String listDirectoryKey = "list_directory";
  static const String gzipKey = "gzip";

  RedirectionConfig get redirections => _redirections;

  num get port => _getFromMap(_config, portKey) ?? 8080;
  String get address => _getFromMap(_config, addressKey) ?? "0.0.0.0";
  bool get isMultithread => _getFromMap(_config, multithreadKey) ?? false;
  String get rootDirectoryPath =>
      _getFromMap(_config, rootDirectoryKey) ?? "./";
  String get logFile => _getFromMap(_config, logFileKey) ?? dartnetLogFile;
  Level get logLevel =>
      logLevels[_getFromMap(_config, logLevelKey)?.toUpperCase()] ?? Level.INFO;
  bool get listDirectory => _getFromMap(_config, listDirectoryKey) ?? false;
  bool get gzip => _getFromMap(_config, gzipKey) ?? true;

  String _correspondingLogString(Level level) => logLevels.keys
      .firstWhere((String key) => logLevels[key] == level, orElse: () => null);

  Map toMap() => {
        addressKey: address,
        portKey: port,
        multithreadKey: isMultithread,
        rootDirectoryKey: rootDirectoryPath,
        logFileKey: logFile,
        logLevelKey: _correspondingLogString(logLevel),
        listDirectoryKey: listDirectory,
        gzipKey: gzip
      };
}

class RedirectionConfig {
  Map _config;
  Map<String, dynamic> _pathConfig;

  static const redirectionsKey = "redirections";
  static const notFoundCodeKey = 404;
  static const pathRedirectionsKey = "path";

  RedirectionConfig(this._config) {
    if (_config != null) {
      _pathConfig = new Map.from(
          _getFromMap(_config, RedirectionConfig.pathRedirectionsKey) ?? {});
    }
  }

  Iterable<String> get paths => _pathConfig.keys;
  String get onNotFound => _getFromMap(_config, notFoundCodeKey);

  Uri operator [](String path) {
    if (_pathConfig.containsKey(path)) {
      if (_pathConfig[path] is String) {
        var redirect = _pathConfig[path];
        while (redirect is String && _pathConfig.containsKey(redirect)) {
          redirect = _pathConfig[redirect];
        }
        _pathConfig[path] = redirect is Uri ? redirect : Uri.parse(redirect);
      }
      return _pathConfig[path];
    }
    return null;
  }
}

class HttpsConfig {
  Map _config;

  static const String httpsKey = "https";
  static const String certKey = "cert";
  static const String keyKey = "key";
  static const String keyPasswordKey = "password_key";

  HttpsConfig(this._config);

  String get certPath => _getFromMap(_config, certKey);
  String get keyPath => _getFromMap(_config, keyKey);
  String get passwordKey => _getFromMap(_config, keyPasswordKey);

  bool get isValid => certPath != null && certKey != null;
}
